src/udx.c

```diff
@@ -578,201 +640,215 @@ mtu_unprobeify_packet (udx_packet_t *pkt, udx_stream_t *stream) {
   debug_printf("\n");
 }
 
-static int
-send_state_packet (udx_stream_t *stream) {
-  if ((stream->status & UDX_STREAM_CONNECTED) == 0) return 0;
+// todo: inefficient
 
-  stream->deferred_ack = 0;
+static udx_stream_t *
+get_stream (udx_socket_t *socket) {
+  for (uint32_t i = 0; i < socket->udx->streams_len; i++) {
+    udx_stream_t *stream = socket->udx->streams[i];
+    if (stream->socket == socket && stream_write_wanted(stream)) {
+      return stream;
+    }
+  }
+  return NULL;
+}
+
+// void
+// udx__initialize_stream_queue (udx_socket_t *socket) {
+//   assert(socket->stream_queue.len == 0);
+//
+//   for (uint32_t i = 0; i < socket->udx->streams_len; i++) {
+//     udx_stream_t *stream = socket->udx->streams[i];
+//     if (stream->socket == socket && stream_write_wanted(stream)) {
+//       udx__fifo_push(&socket->stream_queue, stream);
+//     }
+//   }
+// }
+
+// sending packets
+// while the socket is writable send until EAGAIN or all packets are sent.
+// 1. get a packet to send 'get_packet'. if no packet is available it will return NULL
+// 2. if packet sends 'confirm_packet'
+//    this frees state packets, shrinks the write_t by the amount sent, etc.
+// 3. if packet fails to send call 'undo_packet'
+//    undo packet _must_ be called in reverse order of get_packet
+
+udx_packet_t *
+udx__get_packet (udx_socket_t *socket) {
```

Comment:
when first reading this `get` made me thing this was immut, maybe _shift_packet or something similar to make it appear like a mutation

Reply:
How about `take` and `untake` for udx__cancel_packet?

Reply:
`shift` and `unshift` work too